packages/eslint-plugin/src/rules/thenable-in-promise-aggregators.ts

```diff
@@ -0,0 +1,258 @@
+    function isPromiseLike(program: ts.Program, type: ts.Type): boolean {
+      if (type.isIntersection()) {
+        return type.types.some(t => isPromiseLike(program, t));
+      }
+      if (type.isUnion()) {
+        return type.types.every(t => isPromiseLike(program, t));
+      }
+
+      const symbol = type.getSymbol();
+      if (!symbol) {
+        return false;
+      }
+
+      if (
+        symbol.getName() === 'PromiseConstructor' &&
+        isSymbolFromDefaultLibrary(program, symbol)
+      ) {
+        return true;
+      }
+
+      if (symbol.flags & (ts.SymbolFlags.Class | ts.SymbolFlags.Interface)) {
+        const checker = program.getTypeChecker();
+
+        for (const baseType of checker.getBaseTypes(type as ts.InterfaceType)) {
+          if (isPromiseClassLike(program, baseType)) {
+            return true;
+          }
+        }
+      }
+
+      return false;
+    }
+
+    return {
+      CallExpression(node: TSESTree.CallExpression): void {
+        const callee = skipChainExpression(node.callee);
+        if (callee.type !== AST_NODE_TYPES.MemberExpression) {
+          return;
+        }
+
+        if (callee.computed) {
+          if (
+            callee.property.type !== AST_NODE_TYPES.Literal ||
+            typeof callee.property.value !== 'string' ||
+            !aggregateFunctionNames.includes(callee.property.value)
+          ) {
+            return;
+          }
+        } else if (!aggregateFunctionNames.includes(callee.property.name)) {
+          return;
+        }
+
+        const callerType = services.getTypeAtLocation(callee.object);
+        if (!isPromiseLike(services.program, callerType)) {
+          return;
+        }
+
+        const args = node.arguments;
+        if (args.length < 1) {
+          return;
+        }
+
+        const arg = args[0];
+        if (arg.type === AST_NODE_TYPES.ArrayExpression) {
+          const { elements } = arg;
+          if (elements.length === 0) {
+            return;
+          }
+
+          for (const element of elements) {
+            if (element == null) {
+              context.report({
+                messageId: 'inArray',
+                node: arg,
+              });
+              return;
+            }
+            const elementType = services.getTypeAtLocation(element);
+            if (isTypeAnyType(elementType) || isTypeUnknownType(elementType)) {
+              continue;
+            }
+
+            const originalNode = services.esTreeNodeToTSNodeMap.get(element);
+            if (tsutils.isThenableType(checker, originalNode, elementType)) {
+              continue;
+            }
+
+            context.report({
+              messageId: 'inArray',
+              node: element,
+            });
+          }
+          return;
+        }
+
+        const argType = services.getTypeAtLocation(arg);
+        if (isTypeAnyType(argType) || isTypeUnknownType(argType)) {
+          return;
+        }
+
+        if (checker.isArrayType(argType)) {
```

Comment:
I noticed that we don't check for unions/intersections here, so the following code produces lint errors

```ts
declare const a: Promise<string>[] | Promise<number>[]
Promise.all(a)

declare const b: [Promise<number>] | [Promise<string>]
Promise.all(b)

declare const c: Promise<string>[] & Promise<number>[]
Promise.all(c)

declare const d: [Promise<number>] & [Promise<string>]
Promise.all(d)
```

[playground](https://deploy-preview-8094--typescript-eslint.netlify.app/play/#ts=5.3.2&showAST=es&fileType=.tsx&code=CYUwxgNghgTiAEYD2A7AzgF3lAXPACjEgLYCWaIAPJjKSgOYB8A2gLrwA%2BBRZFlKAV2IAjEDBasAUIRLkQAOigQIACigBKSZNCRYCZOizC8zGbyqCRYxuy6mec6hloMb0hxUXKVwzdvDQcIiomIh4Zo40dExs8ABk3LJ8lqLibO5JCkqqYH46gfohWMAmEclCqTbx8PaZTi5MUmVZ3sDqQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6fAC0SYEMAjJLWa1i0APYBbSskS0eAc3nRE8nvjHRUGSImjjokcGAC%2BIY0A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

Reply:
Added lots of tests around these, including absolutely psychotic stuff like the following:

```ts
type Foo = { new (): [number | Promise<number>] };
declare const foo: Foo;
class Baz extends foo {}
declare const baz: Baz;
await Promise.all(baz);
```